Add favicon to internal routes `/favicon.ico`

hahmed opened the pull request:
### Summary

When creating a new app and running the rails application (rails welcome page), there is an error with the `/favicon.ico` request. This PR adds an internal route for the rails template to serve the favicon by sending back an `head :ok` response.

Previously the error we got:

```
Started GET "/favicon.ico" for 127.0.0.1 at 2017-06-21 10:10:01 +0100

ActionController::RoutingError (No route matches [GET] "/favicon.ico"):
```

I have moved the favicon into a header element -- one less route to override.

### Other Information

Fixes https://github.com/rails/rails/issues/29578

To replicate the error, create a new rails app, on chrome run `rails s` then hard refresh should make a request for the favicon.

Screenshot:

<img width="1440" alt="Screenshot 2021-07-24 at 21 21 07" src="https://user-images.githubusercontent.com/658005/126880378-d4bb307c-fe47-4155-bae9-bc98e4c061ad.png">

hahmed committed:
Add favicon to header dom element for welcome controller.

Co-authored-by: Sunny Ripert <sunny@sunfox.org>
Co-authored-by: Petrik de Heus <petrik@deheus.net>

p8 commented on railties/lib/rails/application/finisher.rb:
```diff
@@ -175,6 +175,7 @@ def self.complete(_state)
 
           routes_reloader.run_after_load_paths = -> do
             app.routes.append do
+              get "/favicon.ico" => "rails/welcome#favicon", internal: true
```
We could also let the route handle the 204 and leave out the changes to the controller.
```suggestion
              get "/favicon.ico" => lambda { |_| [204, {}, []] }, internal: true
```
Or even:
```suggestion
              get "/favicon.ico" => lambda { |_| [200, {}, File.new('path/to/favicon')] } internal: true
```

p8 replied:
If would be nice if we could just do:
```
get "/favicon.ico" => :head, internal: true`
```

p8 replied:
Looking at the guides, they use the following syntax:
```
get "/favicon.ico", to: ->(env) { [204, {}, ['']] }, internal: true
```

hahmed replied:
I considered using `get "/favicon.ico", to: ->(env) { [204, {}, ['']] }, internal: true`.

I ended up going with defining it on the controller, similar to the welcome controller because a new user is one day going to start digging into how the welcome controller is displayed and they will see it's just a basic route 😀

Consistency here is nice, but I'm more than happy to change it.

p8 replied:
The inline env has my preference as it is about as long as the "controller action" implementation, without an extra step of indirection.
But I don't have a strong opinion about it 😄

p8 commented:
The root route to the welcomes_controller gets overridden in apps.
Does this also work as expected for the 'favicon.ico'?
Will it be confusing for new users why the favicon.ico returns a 204 by default?

p8 commented:
An older but closed PR that tried adding the favicon to `/public` https://github.com/rails/rails/pull/29585

p8 commented:
We could also just inline the icon in the welcome template.
```
<link rel="icon" href="data:image/svg+xml,%3Csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2016%2016'%3E%3Ctext%20x='0'%20y='14'%20style='filter:%20invert(100%);'%3E🦄%3C/text%3E%3C/svg%3E" type="image/svg+xml" />
```
https://stackoverflow.com/a/62438464

hahmed commented:
> We could also just inline the icon in the welcome template.

@p8 inlining the favicon works for me 🙂

zzak commented:
@hahmed Curious where you got the svg from, does it show up in the screenshot in the description of this PR too?

LGTM though, I generally like this approach but we may want to move the SVG content to a helper that can be tested, this way if Rails ever changes its logo a code search for "logo" will show up. 🤔

sunny commented:
> Curious where you got the svg from 

It’s actually a jpg. Which is not ideal since it has no transparency:

![image](https://user-images.githubusercontent.com/132/125191392-7a269580-e242-11eb-83a2-4d309104ff8a.png)

An SVG would be great here. Using [this SVG](https://commons.wikimedia.org/wiki/File:Ruby_on_Rails_logo.svg) it would look more like this:

![image](https://user-images.githubusercontent.com/132/125191531-3d0ed300-e243-11eb-8e0a-6e19b7237128.png)

> we may want to move the SVG content to a helper that can be tested, this way if Rails ever changes its logo a code search for "logo" will show up.

If we do that for the favicon it would mean doing this for all other images on the page (which also include the logo). Perhaps a comment, e.g. `<!-- Rails logo -->` would be enough, here?

sunny commented on railties/lib/rails/templates/rails/welcome/index.html.erb:
```diff
@@ -4,6 +4,8 @@
   <title>Ruby on Rails</title>
   <meta charset="utf-8">
   <meta name="viewport" content="width=device-width">
+  <link rel="shortcut icon" href="data:image/jpg;base64,/9j/4AAQSkZJRgABAQAASABIAAD/4QBYRXhpZgAATU0AKgAAAAgAAgESAAMAAAABAAEAAIdpAAQAAAABAAAAJgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAABz6ADAAQAAAABAAACWAAAAAD/7QA4UGhvdG9zaG9wIDMuMAA4QklNBAQAAAAAAAA4QklNBCUAAAAAABDUHYzZjwCyBOmACZjs+EJ+/8AAEQgCWAHPAwEiAAIRAQMRAf/EAB8AAAEFAQEBAQEBAAAAAAAAAAABAgMEBQYHCAkKC//EALUQAAIBAwMCBAMFBQQEAAABfQECAwAEEQUSITFBBhNRYQc ...
```
Here is an SVG version of the logo for better transparency support. It also ends up being smaller.

I picked [this SVG from wikimedia](https://commons.wikimedia.org/wiki/File:Ruby_on_Rails_logo.svg), compressed and url-encoded:

```suggestion
  <!-- Ruby on Rails logo as a favicon -->
  <link rel="shortcut icon" href="data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='259.971' height='330'%3E%3Cpath d='M31.845 0C16.319-.07-.155 15.28 0 31.438V330h228.125c15.526.07 32-15.31 31.844-31.468V0H31.845zm6.469 5.906h214.968v284.969c.147 15.424-15.401 30.097-30.03 30.032H8.281V35.938C8.136 20.515 23.684 5.84 38.314 5.906z'/%3E%3Cpath d='M38.282 5.75C23.652 5.683 8.074 20.358 8.22 35.781v157.97l26.719-45.157-25-62L62.548 5.906 38.282 5.75zM8.22 202.625l.047 118.297h53.508L8.22 202.622zm245.058 1.344l-47.964 116.937h17.928c14.63.065 30.178-14.607 30.031-30.03l.005-86.905zm-122.406 46.647l-62.652 70.29 136.234.016-73.582-70.307z' fill='%238b1e25'/%3E%3Cpath d='M62.568 5.876l.719 73.016-53.38 7.68L34.6 147.86 63.88 78.548l73-28L62.568 5.876zm-28.469 144.14L8.193 193.799l.156 5.563L65.88 216.46l-.859 104.432h3.219l62.64-70.25 73.563 70.297.875-.031 47.96-116.93v-3.562L130.88 249.548l-64.994-33.105-31.787-66.426z' fill='%23a42f38'/%3E%3Cpath d='M62.574 5.862L9.892 86.59l53.423-7.702-.74-73.026zM8.262 199.237v3.406H8.23l53.5 118.25 3.325.013.832-104.45-57.625-17.219z' fill='%23932630'/%3E%3Cpath d='M231.985 149.75l-29.008-71.989-71.492-27.51 72.438-44.313-.957 71.82 50.322 15.576-.014 4.408-20.788 52.008 20.796 50.668-50.8 20.392 29.503-71.06z' fill='%23c68a66'/%3E%3Cpath d='M131.58 50.17l72.343-44.22-141.125-.03 68.781 44.25zm71.374 27.562l50.335 15.607-49.366-87.42-.969 71.813zm50.313 20l-21 52.093 21 50.594V97.732z' fill='%23be785f'/%3E%3Cpath d='M253.423 5.42l-49.5.5 49.358 87.42.142-87.92zM130.485 49.262l-66.5 28.5-31.5 71.5 33.39 67.202 64.995 33.121 71.62-28.776 29.62-71.078-29.125-71.969-72.5-28.5z' fill='%23aa4042'/%3E%3Cpath d='M163.315 16.007c-1.165.02-1.982.16-2.187.375-.41.43-.05 2.459.78 4.5 1.353 3.323 1.928 3.753 5.688 4.094 2.31.209 4.746.434 5.407.53.66.097 1.494-.28 1.843-.843.35-.563.113-2.527-.5-4.375-1.044-3.152-1.448-3.407-6.78-4.031-1.564-.183-3.085-.271-4.25-.25zm-41.187 7.687c-2.216 0-9.361 3.986-9.97 5.563-.684 1.777 6.218 10.5 8.313 10.5.807 0 3.263-1.203 5.47-2.656 2.206-1.454 4.03-3.158 4.03-3.813 0-1.842-6.341-9.594-7.843-9.594zm94.406 7.032c-2.425 0-.296 3.841 3.437 6.187 4.508 2.833 6.782 3.45 6.782 1.875 0-1.07-8.865-8.062-10.22-8.062zm-45.344.343c-3.42.084-6.765.594-12.375 1.594-11.768 2.098-26.145 6.65-31.031 9.844-18.519 12.102-34.553 29.989-49.125 54.812-20.779 35.396-32.212 76.742-35.969 129.907l-.656 9.156c.11.054.247.107.406.156h51.938c28.804 0 52.585-.211 52.843-.469.26-.258-1.447-6.254-3.812-13.312-7.665-22.878-12.576-49.591-12.625-68.563-.065-24.924 7.658-52.563 19.75-70.656 6.497-9.722 16.57-19.692 19.937-19.719 1.29-.01 4.347-1.367 6.813-3.031 2.466-1.664 7.652-3.993 11.5-5.156 11.917-3.604 27.986-1.552 42.844 5.437 5.25 2.47 6.597 2.78 7.312 1.657 1.821-2.863.748-4.987-5-9.75-15.186-12.584-31.614-19.212-52.719-21.313-3.57-.355-6.166-.572-8.562-.594-.5-.01-.98-.012-1.469 0zM76.503 59.507c-2.25.415-9.753 10.12-8.47 11.281 2.755 2.487 11.984 7.965 12.688 7.531.454-.279 2.173-2.488 3.844-4.906 1.671-2.418 3.031-4.815 3.031-5.312 0-.773-6.666-6.074-10.718-8.532a.522.522 0 00-.375-.062zm139.718 1.687c-.8 0-1.49.148-1.843.5-.553.55-.722 1.758-.375 2.657.823 2.136 8.218 1.742 8.218-.438 0-1.355-3.597-2.72-6-2.719zm-29.968 1c-.417.01-.935.1-1.563.25-3.655.87-6.344 2.26-6.344 3.282 0 .447.923 1.625 2.032 2.625 1.109.999 2 2.444 2 3.187 0 1.638.79 1.679 4.906.25 3.619-1.256 3.711-1.677 1.687-6.5-.987-2.352-1.468-3.114-2.718-3.094zm-25.75 19.719c-1.242 0-4.344 3.339-4.344 4.688 0 1.732 5.047 9.375 6.187 9.375.469 0 1.567-1.129 2.47-2.5 1.557-2.368 1.501-2.727-1.032-7.032-1.468-2.493-2.932-4.531-3.281-4.531zM41.878 115.038c-1.356 0-5.688 10.408-5.688 13.656 0 1.21 11.088 6.163 14.219 6.344 1.137.066 2.297-1.902 3.844-6.594 1.797-5.452 1.94-6.913.875-7.875-1.673-1.508-11.298-5.531-13.25-5.531zm101.906.406c-.293.058-.53.277-.75.625-.524.825-.96 2.011-.969 2.657-.008.645-.266 1.869-.594 2.718-.385 1 .948 2.974 3.782 5.594 4.903 4.534 5.413 4.722 6.28 2.469 1.533-3.976.543-7.101-3.468-11-2.101-2.043-3.224-2.951-3.969-3.063a.91.91 0 00-.312 0zm-5.313 38.938c-1.138-.045-1.446.835-1.437 2.594.028 5.62 1.033 7.032 6.875 9.625 7.67 3.403 8.311 3.36 7.625-.594-.928-5.347-1.098-5.576-6.656-8.75-3.307-1.888-5.268-2.83-6.407-2.875zM21.815 183.319c-1.694 0-6.38 15.15-5.437 17.594.508 1.319 4.266 1.95 13.843 2.313l4.625.156 1.844-8.063c1.11-4.842 1.495-8.356.938-8.843-.904-.792-12.74-3.157-15.813-3.157zm124.063 20.375c-.301.011-.496.058-.563.125-.332.331.177 3.067 1.125 6.063 1.459 4.606 2.15 5.492 4.469 5.844 3.856.583 14.344.816 14.344.312 0-2.093-4.282-9.56-5.72-9.969-4.276-1.218-11.549-2.449-13.655-2.375zm73.468 49.125c-2.396 0-4.876.084-6.843.282-5.788.578-7.734 1.261-10.25 3.5-2.751 2.447-3.148 3.488-3.47 9.156-.291 5.15.023 7.044 1.688 9.687 2.578 4.09 5.547 5.262 13.5 5.282 7.552.018 9.03.79 8.594 4.593l-.344 2.907-11.343.28-11.344.282V298.757h12.656c9.339 0 13.411-.376 15.594-1.5 4.16-2.142 6.031-6.343 6.031-13.594 0-5.683-.282-6.452-3.437-9.594-3.474-3.458-3.916-3.573-14.72-3.843-4.38-.11-6.03-1.11-6.03-3.657 0-2.831 3.807-3.937 13.656-3.937h8.5v-8.844l-6.063-.687c-1.668-.193-3.979-.284-6.375-.282zm-76.28.657l-5.282.312-5.313.313-.28 21.062c-.151 11.582-.022 21.608.25 22.313.309.803 2.378 1.28 5.562 1.28h5.062v-45.28zm-112.938.125v45.156h11l.312-5.75c.18-3.416.738-5.782 1.406-5.781.622 0 3.278 2.606 5.875 5.78l4.72 5.75h6c3.296 0 6-.244 6-.562 0-.318-2.068-3.208-4.563-6.437-2.496-3.23-4.532-6.204-4.532-6.594 0-.39 2.036-2.592 4.532-4.906l4.5-4.188c.016-.054.046-.101.062-.156v-7.687c0-7.545-.126-7.922-3.437-11.22l-3.438-3.405H30.128zm68.687 0c-16.78.035-17.25.782-17.25 26.53v18.626h5.031c5.013 0 5.048-.01 5.063-3.25.025-5.213.975-5.904 7.625-5.563l5.969.282.312 4.343.313 4.313 5.218-.313 5.25-.312.313-16.563c.484-26.232-.725-28.13-17.844-28.093zm60.375 0v45.218l14.875-.281 14.875-.281V288.226l-9.062-.5-9.094-.5-.281-16.813-.25-16.812H159.19zM48.628 263.63c1.575.022 3.107.237 4.125.626 2.253.86 2.253 9.951 0 10.812-2.714 1.037-9.04.674-10.313-.593-.665-.663-1.219-2.825-1.219-4.813 0-1.988.554-4.15 1.22-4.812.795-.793 3.562-1.254 6.187-1.22zm50.156.032c2.474 0 4.985.425 5.75 1.187.665.663 1.219 3.497 1.219 6.313s-.554 5.65-1.22 6.312c-1.829 1.822-9.414 1.515-11.405-.469-2.05-2.04-1.575-11.062.656-12.468.773-.488 2.276-.762 3.937-.844.347-.017.71-.031 1.063-.031z' fill='%23f2f2f2'/%3E%3C/svg%3E" />
```

hahmed commented:
> LGTM though, I generally like this approach but we may want to move the SVG content to a helper that can be tested, this way if Rails ever changes its logo a code search for "logo" will show up. 🤔

I have moved this to a var in the view, added a descriptive name, hope that's ok? 

Could you clarify how it could be tested? Also I'm not sure we are testing the other 2 images on the page either...

sunny commented on railties/lib/rails/templates/rails/welcome/index.html.erb:
```diff
@@ -1,9 +1,12 @@
+<% ruby_on_rails_logo_favicon_base64 = "data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='259.971' height='330'%3E%3Cpath d='M31.845 0C16.319-.07-.155 15.28 0 31.438V330h228.125c15.526.07 32-15.31 31.844-31.468V0H31.845zm6.469 5.906h214.968v284.969c.147 15.424-15.401 30.097-30.03 30.032H8.281V35.938C8.136 20.515 23.684 5.84 38.314 5.906z'/%3E%3Cpath d='M38.282 5.75C23.652 5.683  ...
```
It’s actually not base64-encoded, simply url-encoded, here:

```suggestion
<% ruby_on_rails_logo_favicon_data_uri = "data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='259.971' height='330'%3E%3Cpath d='M31.845 0C16.319-.07-.155 15.28 0 31.438V330h228.125c15.526.07 32-15.31 31.844-31.468V0H31.845zm6.469 5.906h214.968v284.969c.147 15.424-15.401 30.097-30.03 30.032H8.281V35.938C8.136 20.515 23.684 5.84 38.314 5.906z'/%3E%3Cpath d='M38.282 5.75C23.652 5.683 8.074 20.358 8.22 35.781v157.97l26.719-45.157-25-62L62.548 5.906 38.282 5.75zM8.22 202.625l.047 118.297h53.508L8.22 202.622zm245.058 1.344l-47.964 116.937h17.928c14.63.065 30.178-14.607 30.031-30.03l.005-86.905zm-122.406 46.647l-62.652 70.29 136.234.016-73.582-70.307z' fill='%238b1e25'/%3E%3Cpath d='M62.568 5.876l.719 73.016-53.38 7.68L34.6 147.86 63.88 78.548l73-28L62.568 5.876zm-28.469 144.14L8.193 193.799l.156 5.563L65.88 216.46l-.859 104.432h3.219l62.64-70.25 73.563 70.297.875-.031 47.96-116.93v-3.562L130.88 249.548l-64.994-33.105-31.787-66.426z' fill='%23a42f38'/%3E%3Cpath d='M62.574 5.862L9.892 86.59l53.423-7.702-.74-73.026zM8.262 199.237v3.406H8.23l53.5 118.25 3.325.013.832-104.45-57.625-17.219z' fill='%23932630'/%3E%3Cpath d='M231.985 149.75l-29.008-71.989-71.492-27.51 72.438-44.313-.957 71.82 50.322 15.576-.014 4.408-20.788 52.008 20.796 50.668-50.8 20.392 29.503-71.06z' fill='%23c68a66'/%3E%3Cpath d='M131.58 50.17l72.343-44.22-141.125-.03 68.781 44.25zm71.374 27.562l50.335 15.607-49.366-87.42-.969 71.813zm50.313 20l-21 52.093 21 50.594V97.732z' fill='%23be785f'/%3E%3Cpath d='M253.423 5.42l-49.5.5 49.358 87.42.142-87.92zM130.485 49.262l-66.5 28.5-31.5 71.5 33.39 67.202 64.995 33.121 71.62-28.776 29.62-71.078-29.125-71.969-72.5-28.5z' fill='%23aa4042'/%3E%3Cpath d='M163.315 16.007c-1.165.02-1.982.16-2.187.375-.41.43-.05 2.459.78 4.5 1.353 3.323 1.928 3.753 5.688 4.094 2.31.209 4.746.434 5.407.53.66.097 1.494-.28 1.843-.843.35-.563.113-2.527-.5-4.375-1.044-3.152-1.448-3.407-6.78-4.031-1.564-.183-3.085-.271-4.25-.25zm-41.187 7.687c-2.216 0-9.361 3.986-9.97 5.563-.684 1.777 6.218 10.5 8.313 10.5.807 0 3.263-1.203 5.47-2.656 2.206-1.454 4.03-3.158 4.03-3.813 0-1.842-6.341-9.594-7.843-9.594zm94.406 7.032c-2.425 0-.296 3.841 3.437 6.187 4.508 2.833 6.782 3.45 6.782 1.875 0-1.07-8.865-8.062-10.22-8.062zm-45.344.343c-3.42.084-6.765.594-12.375 1.594-11.768 2.098-26.145 6.65-31.031 9.844-18.519 12.102-34.553 29.989-49.125 54.812-20.779 35.396-32.212 76.742-35.969 129.907l-.656 9.156c.11.054.247.107.406.156h51.938c28.804 0 52.585-.211 52.843-.469.26-.258-1.447-6.254-3.812-13.312-7.665-22.878-12.576-49.591-12.625-68.563-.065-24.924 7.658-52.563 19.75-70.656 6.497-9.722 16.57-19.692 19.937-19.719 1.29-.01 4.347-1.367 6.813-3.031 2.466-1.664 7.652-3.993 11.5-5.156 11.917-3.604 27.986-1.552 42.844 5.437 5.25 2.47 6.597 2.78 7.312 1.657 1.821-2.863.748-4.987-5-9.75-15.186-12.584-31.614-19.212-52.719-21.313-3.57-.355-6.166-.572-8.562-.594-.5-.01-.98-.012-1.469 0zM76.503 59.507c-2.25.415-9.753 10.12-8.47 11.281 2.755 2.487 11.984 7.965 12.688 7.531.454-.279 2.173-2.488 3.844-4.906 1.671-2.418 3.031-4.815 3.031-5.312 0-.773-6.666-6.074-10.718-8.532a.522.522 0 00-.375-.062zm139.718 1.687c-.8 0-1.49.148-1.843.5-.553.55-.722 1.758-.375 2.657.823 2.136 8.218 1.742 8.218-.438 0-1.355-3.597-2.72-6-2.719zm-29.968 1c-.417.01-.935.1-1.563.25-3.655.87-6.344 2.26-6.344 3.282 0 .447.923 1.625 2.032 2.625 1.109.999 2 2.444 2 3.187 0 1.638.79 1.679 4.906.25 3.619-1.256 3.711-1.677 1.687-6.5-.987-2.352-1.468-3.114-2.718-3.094zm-25.75 19.719c-1.242 0-4.344 3.339-4.344 4.688 0 1.732 5.047 9.375 6.187 9.375.469 0 1.567-1.129 2.47-2.5 1.557-2.368 1.501-2.727-1.032-7.032-1.468-2.493-2.932-4.531-3.281-4.531zM41.878 115.038c-1.356 0-5.688 10.408-5.688 13.656 0 1.21 11.088 6.163 14.219 6.344 1.137.066 2.297-1.902 3.844-6.594 1.797-5.452 1.94-6.913.875-7.875-1.673-1.508-11.298-5.531-13.25-5.531zm101.906.406c-.293.058-.53.277-.75.625-.524.825-.96 2.011-.969 2.657-.008.645-.266 1.869-.594 2.718-.385 1 .948 2.974 3.782 5.594 4.903 4.534 5.413 4.722 6.28 2.469 1.533-3.976.543-7.101-3.468-11-2.101-2.043-3.224-2.951-3.969-3.063a.91.91 0 00-.312 0zm-5.313 38.938c-1.138-.045-1.446.835-1.437 2.594.028 5.62 1.033 7.032 6.875 9.625 7.67 3.403 8.311 3.36 7.625-.594-.928-5.347-1.098-5.576-6.656-8.75-3.307-1.888-5.268-2.83-6.407-2.875zM21.815 183.319c-1.694 0-6.38 15.15-5.437 17.594.508 1.319 4.266 1.95 13.843 2.313l4.625.156 1.844-8.063c1.11-4.842 1.495-8.356.938-8.843-.904-.792-12.74-3.157-15.813-3.157zm124.063 20.375c-.301.011-.496.058-.563.125-.332.331.177 3.067 1.125 6.063 1.459 4.606 2.15 5.492 4.469 5.844 3.856.583 14.344.816 14.344.312 0-2.093-4.282-9.56-5.72-9.969-4.276-1.218-11.549-2.449-13.655-2.375zm73.468 49.125c-2.396 0-4.876.084-6.843.282-5.788.578-7.734 1.261-10.25 3.5-2.751 2.447-3.148 3.488-3.47 9.156-.291 5.15.023 7.044 1.688 9.687 2.578 4.09 5.547 5.262 13.5 5.282 7.552.018 9.03.79 8.594 4.593l-.344 2.907-11.343.28-11.344.282V298.757h12.656c9.339 0 13.411-.376 15.594-1.5 4.16-2.142 6.031-6.343 6.031-13.594 0-5.683-.282-6.452-3.437-9.594-3.474-3.458-3.916-3.573-14.72-3.843-4.38-.11-6.03-1.11-6.03-3.657 0-2.831 3.807-3.937 13.656-3.937h8.5v-8.844l-6.063-.687c-1.668-.193-3.979-.284-6.375-.282zm-76.28.657l-5.282.312-5.313.313-.28 21.062c-.151 11.582-.022 21.608.25 22.313.309.803 2.378 1.28 5.562 1.28h5.062v-45.28zm-112.938.125v45.156h11l.312-5.75c.18-3.416.738-5.782 1.406-5.781.622 0 3.278 2.606 5.875 5.78l4.72 5.75h6c3.296 0 6-.244 6-.562 0-.318-2.068-3.208-4.563-6.437-2.496-3.23-4.532-6.204-4.532-6.594 0-.39 2.036-2.592 4.532-4.906l4.5-4.188c.016-.054.046-.101.062-.156v-7.687c0-7.545-.126-7.922-3.437-11.22l-3.438-3.405H30.128zm68.687 0c-16.78.035-17.25.782-17.25 26.53v18.626h5.031c5.013 0 5.048-.01 5.063-3.25.025-5.213.975-5.904 7.625-5.563l5.969.282.312 4.343.313 4.313 5.218-.313 5.25-.312.313-16.563c.484-26.232-.725-28.13-17.844-28.093zm60.375 0v45.218l14.875-.281 14.875-.281V288.226l-9.062-.5-9.094-.5-.281-16.813-.25-16.812H159.19zM48.628 263.63c1.575.022 3.107.237 4.125.626 2.253.86 2.253 9.951 0 10.812-2.714 1.037-9.04.674-10.313-.593-.665-.663-1.219-2.825-1.219-4.813 0-1.988.554-4.15 1.22-4.812.795-.793 3.562-1.254 6.187-1.22zm50.156.032c2.474 0 4.985.425 5.75 1.187.665.663 1.219 3.497 1.219 6.313s-.554 5.65-1.22 6.312c-1.829 1.822-9.414 1.515-11.405-.469-2.05-2.04-1.575-11.062.656-12.468.773-.488 2.276-.762 3.937-.844.347-.017.71-.031 1.063-.031z' fill='%23f2f2f2'/%3E%3C/svg%3E" %>
```

hahmed replied:
Thanks @sunny 👍

sunny commented on railties/lib/rails/templates/rails/welcome/index.html.erb:
```diff
@@ -1,9 +1,12 @@
   <title>Ruby on Rails</title>
   <meta charset="utf-8">
   <meta name="viewport" content="width=device-width">
+  <link rel="shortcut icon" href="<%= ruby_on_rails_logo_favicon_base64 %>" />
```
(Oops, also applies here, sorry I didn’t include that as a suggestion as well 😅)

```suggestion
  <link rel="shortcut icon" href="<%= ruby_on_rails_logo_favicon_data_uri %>" />
```

p8 commented:
The logo is a bit outdated. I'll try to see if I can generate an SVG from https://www.pngegg.com/en/png-zrbae

p8 commented:
I've created an SVG that looks like this:
<img width="80" alt="image" src="https://user-images.githubusercontent.com/28561/126079502-7d93be54-f3aa-464b-8352-50e3f855e9fb.png">

```svg
<?xml version="1.0" encoding="UTF-8"?>
<svg width="320px" height="320px" viewBox="0 0 320 320" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <title>Icon</title>
    <g id="Icon" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Ruby-On-Rails-Logo" transform="translate(5.000000, 20.000000)" fill="#D81E00">
            <path d="M21.9330706,285 L168.175426,285 C147.940858,150.691706 196.882383,66.3142902 315,31.8677536 C315,21.7851284 315,31.8677536 315,21.7851284 C161.25283,24.9116278 63.5638533,112.649918 21.9330706,285 Z" id="Path"></path>
            <polygon id="Path" points="40.4082643 185.91571 12.7326494 174.401663 -1.42108547e-14 205.44271 29.5711966 216.647115"></polygon>
            <polygon id="Path" points="180.473102 268.073643 204.936016 276.410966 204.936016 254.508673 180.489644 244.338107"></polygon>
            <polygon id="Path" points="100.589519 97.2287606 76.6840156 79.1687021 55.2440791 100.183105 81.5143239 117.433725"></polygon>
            <polygon id="Path" points="184.575679 184.896295 207.95203 200.24661 211.727729 181.205266 189.863653 164.68746"></polygon>
            <polygon id="Path" points="261.73401 65.8959446 269.354234 82.32954 285.318256 72.7048596 278.310129 56.9829895"></polygon>
            <polygon id="Path" points="261.9137 16.1845473 255.598479 7.10542736e-15 232.79602 3.48896323 240.463873 20.0252725"></polygon>
            <polygon id="Path" points="211.934119 111.580675 226.722547 127.792601 238.023925 113.903715 223.468375 96.5468685"></polygon>
            <polygon id="Path" points="179.686858 38.4952218 164.641931 20.0252725 139.84762 32.5552992 156.601324 50.9216753"></polygon>
        </g>
    </g>
</svg>
```

hahmed commented:
Awesome, thanks @p8 I have updated and added a screenshot of the new svg logo

p8 commented:
Thanks @hahmed and @sunny! This looks PR great to me.

WaKeMaTTa commented:
Why I never though about it? Good job ma'men!
